src/domain/usecases/ValidateOptionSetsUseCase.ts

```diff
@@ -0,0 +1,106 @@
+
+export class ValidateOptionSetsUseCase {
+    constructor(private readonly optionSetRepository: OptionSetRepository) {}
+
+    async execute(options: UseCaseOptions): Async<OptionValidationResult[]> {
+        logger.info("Loading optionSets...");
+        const optionsSets = await this.optionSetRepository.getAll();
+        return this.buildOptionValidationResults(optionsSets, options);
+    }
+
+    private buildOptionValidationResults(
+        optionsSets: OptionSet[],
+        options: UseCaseOptions
+    ): OptionValidationResult[] {
+        const duplicateOptions = this.getDuplicateOptions(optionsSets);
+        const validationResults = this.buildValidationResults(optionsSets, options);
+        return duplicateOptions.concat(validationResults).filter(vr => vr.rules.length > 0);
+    }
+
+    private buildValidationResults(
+        optionSets: OptionSet[],
+        options: UseCaseOptions
+    ): OptionValidationResult[] {
+        const allOptions = optionSets.flatMap(optionSet =>
+            optionSet.options.map((option, optionIndex) => ({ option, optionSet, optionIndex }))
+        );
+
+        return allOptions.map(({ option, optionIndex, optionSet }): OptionValidationResult => {
+            const invalidLengthRules = this.validateLength(option, options.codeLength);
+            const namingConventionRules = this.validateNamingConventions(option);
+            const misplacedCommasRules = this.validateMisplacedCommas(option);
+            const orderRules = this.validateOrder(option, optionIndex);
+            return {
+                option,
+                optionSet,
+                rules: [
+                    ...invalidLengthRules,
+                    ...namingConventionRules,
+                    ...misplacedCommasRules,
+                    ...orderRules,
+                ],
+            };
+        });
+    }
+
+    private validateLength(option: Option, validationLength: number): OptionValidationRule[] {
+        return option.code.length > validationLength
+            ? [{ type: "invalid_length", message: `Code length is ${option.code.length}` }]
+            : [];
+    }
+
+    private validateNamingConventions(option: Option): OptionValidationRule[] {
+        const hasSpaces = /\s/.test(option.code);
+        const hasSpecialCharacters = option.code.match(/[^\w\s]/) ?? [];
+
+        const spacesRule = hasSpaces ? "spaces" : "";
+        const specialRule =
+            hasSpecialCharacters.length > 0 ? `special characters: ${hasSpecialCharacters.join(",")}` : "";
+
+        const message = [spacesRule, specialRule].filter(Boolean).join(" and ");
+
+        return spacesRule || specialRule
+            ? [{ type: "naming_conventions", message: `Code contains ${message}` }]
+            : [];
+    }
+
+    private validateMisplacedCommas(option: Option): OptionValidationRule[] {
+        return option.code.includes(",")
+            ? [{ type: "misplaced_commas", message: `Code has misplaced commas` }]
```

Comment:
maybe it comes from the spec, but it looks weird, the text is "misplaced commas", but the code condition will be true for any comma found in the code, regardless of its place.